source/common/router/config_impl.h

```diff
@@ -937,6 +965,10 @@ class RouteEntryImplBase : public RouteEntryAndRoute,
 
   private:
     const RouteEntryAndRoute* parent_;
+
+    // Keep an copy of the shared pointer to the parent Route. This is needed
+    // to ensure that the parent Route is not destroyed before this entry.
+    const RouteConstSharedPtr owner_;
```

Comment:
When we create a instance of `DynamicRouteEntry`, the instance will returned directly to the active request. The instance will not be stored by the `RouteEntryImplBase`, so there is no circular reference.

The special case is the `WeightedClusterEntry` which is derived from `DynamicRouteEntry` but the instances of `WeightedClusterEntry` will be store in the `RouteEntryImplBase`. So we will keep the `owner_` of `WeightedClusterEntry` be nullptr to avoid circular references.

Reply:
OK I see, thanks. Can you add more comments around all of this? It's kind of confusing to look through. The other ones about the shared part make sense. Thanks for the explanation.

Reply:
get it.